Basic authentication does not work with HA multi-node cluster

**Describe the bug**
Pull queries and multi-node communication does not work or hangs when ksql is configured for HA (2 or more node) and Basic authentication used. 

This is the exception seen in one of the nodes when running `show queries` or a pull query from another node. I can authenticate with each node directly and run pull queries with keys available in that node. But if a run a pull query available in another node, then the internal node communication fails.
```
[ERROR] 2023-03-01 22:05:49,700 [vert.x-eventloop-thread-11] io.vertx.ext.web.RoutingContext ? - Unhandled exception in router
io.confluent.ksql.api.server.KsqlApiException: Unauthorized
	at io.confluent.ksql.api.server.AuthHandlers.selectHandler(AuthHandlers.java:143)
	at io.confluent.ksql.api.server.AuthHandlers.lambda$setupAuthHandlers$2(AuthHandlers.java:63)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.confluent.ksql.api.server.AuthHandlers.pauseHandler(AuthHandlers.java:202)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.confluent.ksql.api.server.LoggingHandler.handle(LoggingHandler.java:135)
	at io.confluent.ksql.api.server.LoggingHandler.handle(LoggingHandler.java:35)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:68)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:67)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:158)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
```

Both nodes are correctly configured with Basic authentication.
```
authentication.method=BASIC
authentication.realm=KsqlServer-Props
authentication.roles=admin,developer,user,ksql-user
authentication.skip.paths=/chc/live,/chc/ready
```

The root cause is on the line of this code https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/DefaultApiSecurityContext.java#L41
```
    String authToken = null;
    if (server.getAuthenticationPlugin().isPresent()) {
      authToken = server.getAuthenticationPlugin().get().getAuthHeader(routingContext);
    }
```

The above condition expects an `authentication.plugin.class` in order to get the `authHeader` of the user making the request to a node. Then that node uses this `authHeader` later for internal node communication. Because no plugin is used for BASIC authentication, then the header is NULL and causes the Unauthorized errors on the other nodes.

What should be fixed?

The `authHeader` should obtain the value from the `RoutingContext` directly as default in case there is no authentication plugin. In fact, the `AuthenticationPlugin` has a default method, but there is no implementation we can use in the KSQL config.
See https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPlugin.java#LL58C3-L60C4
```
  default String getAuthHeader(final RoutingContext routingContext) {
    return routingContext.request().getHeader("Authorization");
  }
```

The fix would look something like this:
```
# Get the authorization header by default
String authToken = routingContext.request().getHeader("Authorization");
if (server.getAuthenticationPlugin().isPresent()) {
  # If plugin exists, then get the header from it
  authToken = server.getAuthenticationPlugin().get().getAuthHeader(routingContext);
}
```

--- a/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/DefaultApiSecurityContext.java
+++ b/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/DefaultApiSecurityContext.java
@@ -38,7 +38,7 @@ public final class DefaultApiSecurityContext implements ApiSecurityContext {
     }
     final ApiUser apiUser = (ApiUser) user;
 
-    String authToken = null;
+    String authToken = routingContext.request().getHeader("Authorization");
     if (server.getAuthenticationPlugin().isPresent()) {
       authToken = server.getAuthenticationPlugin().get().getAuthHeader(routingContext);
     }

--- a/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryFunctionalTest.java
+++ b/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryFunctionalTest.java
@@ -137,9 +137,6 @@ public class PullQueryFunctionalTest {
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "cache.max.bytes.buffering", 10000)
       .withProperty(KsqlConfig.KSQL_QUERY_PULL_TABLE_SCAN_ENABLED, true)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
 
   private static final TestKsqlRestApp REST_APP_1 = TestKsqlRestApp
@@ -153,9 +150,6 @@ public class PullQueryFunctionalTest {
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "cache.max.bytes.buffering", 10000)
       .withProperty(KsqlConfig.KSQL_QUERY_PULL_TABLE_SCAN_ENABLED, true)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
-      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
 
   @ClassRule

--- a/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryFunctionalTest.java
+++ b/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryFunctionalTest.java
@@ -46,7 +46,6 @@ import io.confluent.ksql.test.util.KsqlIdentifierTestUtil;
 import io.confluent.ksql.test.util.KsqlTestFolder;
 import io.confluent.ksql.test.util.TestBasicJaasConfig;
 import io.confluent.ksql.util.KsqlConfig;
-import io.confluent.ksql.util.KsqlRequestConfig;
 import io.confluent.ksql.util.UserDataProvider;
 import io.vertx.core.net.SocketAddress;
 import java.io.IOException;
@@ -138,6 +137,9 @@ public class PullQueryFunctionalTest {
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "cache.max.bytes.buffering", 10000)
       .withProperty(KsqlConfig.KSQL_QUERY_PULL_TABLE_SCAN_ENABLED, true)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
 
   private static final TestKsqlRestApp REST_APP_1 = TestKsqlRestApp
@@ -151,6 +153,9 @@ public class PullQueryFunctionalTest {
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
       .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "cache.max.bytes.buffering", 10000)
       .withProperty(KsqlConfig.KSQL_QUERY_PULL_TABLE_SCAN_ENABLED, true)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
 
   @ClassRule

--- a/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/ShowQueriesMultiNodeFunctionalTest.java
+++ b/ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/ShowQueriesMultiNodeFunctionalTest.java
@@ -23,6 +23,7 @@ import com.google.errorprone.annotations.Immutable;
 import io.confluent.common.utils.IntegrationTest;
 import io.confluent.ksql.integration.IntegrationTestHarness;
 import io.confluent.ksql.integration.Retry;
+import io.confluent.ksql.rest.client.BasicCredentials;
 import io.confluent.ksql.rest.entity.KsqlEntity;
 import io.confluent.ksql.rest.entity.KsqlHostInfoEntity;
 import io.confluent.ksql.rest.entity.Queries;
@@ -34,10 +35,12 @@ import io.confluent.ksql.rest.server.KsqlRestConfig;
 import io.confluent.ksql.rest.server.TestKsqlRestApp;
 import io.confluent.ksql.rest.server.utils.TestUtils;
 import io.confluent.ksql.serde.FormatFactory;
+import io.confluent.ksql.test.util.TestBasicJaasConfig;
 import io.confluent.ksql.util.PageViewDataProvider;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Supplier;
@@ -59,22 +62,40 @@ public class ShowQueriesMultiNodeFunctionalTest {
   private static final int INT_PORT1 = TestUtils.findFreeLocalPort();
   private static final KsqlHostInfoEntity host0 = new KsqlHostInfoEntity("localhost", INT_PORT0);
   private static final KsqlHostInfoEntity host1 = new KsqlHostInfoEntity("localhost", INT_PORT1);
+
+  private static final String PROPS_JAAS_REALM = "KsqlServer-Props";
+  private static final String KSQL_CLUSTER_ID = "ksql-42";
+  private static final String USER_WITH_ACCESS = "harry";
+  private static final String USER_WITH_ACCESS_PWD = "changeme";
+
+  private static final TestBasicJaasConfig JAAS_CONFIG = TestBasicJaasConfig
+      .builder(PROPS_JAAS_REALM)
+      .addUser(USER_WITH_ACCESS, USER_WITH_ACCESS_PWD, KSQL_CLUSTER_ID)
+      .build();
+
   private static final IntegrationTestHarness TEST_HARNESS = IntegrationTestHarness.build();
   private static final TestKsqlRestApp REST_APP_0 = TestKsqlRestApp
       .builder(TEST_HARNESS::kafkaBootstrapServers)
       .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:" + INT_PORT0)
       .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:" + INT_PORT0)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
   private static final TestKsqlRestApp REST_APP_1 = TestKsqlRestApp
       .builder(TEST_HARNESS::kafkaBootstrapServers)
       .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:" + INT_PORT1)
       .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:" + INT_PORT1)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_METHOD_CONFIG, KsqlRestConfig.AUTHENTICATION_METHOD_BASIC)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
+      .withProperty(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
       .build();
 
   @ClassRule
   public static final RuleChain CHAIN = RuleChain
       .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
       .around(TEST_HARNESS)
+      .around(JAAS_CONFIG)
       .around(REST_APP_0)
       .around(REST_APP_1);
 
@@ -82,10 +103,11 @@ public class ShowQueriesMultiNodeFunctionalTest {
   public static void setUpClass() throws InterruptedException {
     TEST_HARNESS.ensureTopics(2, PAGE_VIEW_TOPIC);
     TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER, FormatFactory.KAFKA, FormatFactory.JSON);
-    RestIntegrationTestUtil.createStream(REST_APP_0, PAGE_VIEWS_PROVIDER);
+    RestIntegrationTestUtil.createStream(REST_APP_0, PAGE_VIEWS_PROVIDER, validCreds());
     RestIntegrationTestUtil.makeKsqlRequest(
         REST_APP_0,
-        "CREATE STREAM S AS SELECT * FROM " + PAGE_VIEW_STREAM + ";"
+        "CREATE STREAM S AS SELECT * FROM " + PAGE_VIEW_STREAM + ";",
+        validCreds()
     );
   }
 
@@ -103,7 +125,8 @@ public class ShowQueriesMultiNodeFunctionalTest {
   private static String getShowQueriesResult(final TestKsqlRestApp restApp) {
     final List<KsqlEntity> results = RestIntegrationTestUtil.makeKsqlRequest(
         restApp,
-        "Show Queries;"
+        "Show Queries;",
+        validCreds()
     );
 
     if (results.size() != 1) {
@@ -153,7 +176,8 @@ public class ShowQueriesMultiNodeFunctionalTest {
   private static QueryExtendedResults getShowQueriesExtendedResult(final TestKsqlRestApp restApp) {
     final List<KsqlEntity> results = RestIntegrationTestUtil.makeKsqlRequest(
         restApp,
-        "Show Queries Extended;"
+        "Show Queries Extended;",
+        validCreds()
     );
 
     if (results.size() != 1) {
@@ -203,5 +227,12 @@ public class ShowQueriesMultiNodeFunctionalTest {
     Set<StreamsTaskMetadata> getTasksMetadata() {
       return tasksMetadata;
     }
+  }
+
+  private static Optional<BasicCredentials> validCreds() {
+    return Optional.of(BasicCredentials.of(
+        USER_WITH_ACCESS,
+        USER_WITH_ACCESS_PWD
+    ));
   }
 }

